sync/syncer.go

```diff
@@ -189,98 +207,188 @@ func (s *Syncer) getLayerFromNeighbors(currenSyncLayer types.LayerID) (*types.La
+		//traverse mesh and check local block syntactic validity return true
+		s.Info("traversing %v view for validation", blk.ID())
+		if bks := vq.traverse(s); len(bks) > 0 {
+			s.Warning(fmt.Sprintf("could not validate %v blocks in block %v view ", len(bks) > 0, blk.ID()))
+			return false
+		}
 	}
 
-	return blocksArr, nil
+	return true
 }
 
-func (s *Syncer) syncMissingContent(blk *types.Block) error {
+func (s *Syncer) DataAvailabilty(blk *types.Block) ([]*types.SerializableTransaction, []*types.ActivationTx, error) {
 	var txs []*types.SerializableTransaction
 	var txerr error
 	wg := sync.WaitGroup{}
 	wg.Add(2)
 	go func() {
 		//sync Transactions
-		txs, txerr = s.Txs(blk)
+		txs, txerr = s.syncTxs(blk.TxIds)
 		wg.Done()
 	}()
 
 	var atxs []*types.ActivationTx
-	//var associated *types.ActivationTx
 	var atxerr error
 
+	//sync ATxs
 	go func() {
-		//sync ATxs
-		atxs, _, atxerr = s.ATXs(blk)
+		atxs, atxerr = s.syncAtxs(blk.AtxIds)
 		wg.Done()
 	}()
 
 	wg.Wait()
 
 	if txerr != nil {
 		s.Warning(fmt.Sprintf("failed fetching block %v transactions %v", blk.ID(), txerr))
-		return txerr
+		return txs, atxs, txerr
 	}
 
 	if atxerr != nil {
 		s.Warning(fmt.Sprintf("failed fetching block %v activation transactions %v", blk.ID(), atxerr))
-		return atxerr
+		return txs, atxs, atxerr
 	}
 
 	s.Info(fmt.Sprintf("fetched all txs for block %v", blk.ID()))
 	s.Info(fmt.Sprintf("fetched all atxs for block %v", blk.ID()))
 
-	if err := s.AddBlockWithTxs(blk, txs, atxs); err != nil {
-		s.Warning(fmt.Sprintf("failed fetching block %v activation transactions %v", blk.ID(), err))
-		return err
+	return txs, atxs, nil
+}
+
+func (s *Syncer) fetchLayerBlockIds(m map[string]p2p.Peer, lyr types.LayerID) ([]types.BlockID, error) {
+	//send request to different users according to returned hashes
+	v := make([]p2p.Peer, 0, len(m))
+	for _, value := range m {
+		v = append(v, value)
+	}
+
+	wrk, output := NewPeersWorker(s, v, &sync.Once{}, LayerIdsReqFactory(lyr))
+	go wrk.Work()
+
+	out := <-output
+	if out == nil {
+		return nil, errors.New("could not get layer ids from any peer")
 	}
 
-	s.Info(fmt.Sprintf("added block with txs to database %v", blk.ID()))
-	return nil
+	idSet := make(map[types.BlockID]struct{}, s.LayerSize)
+	ids := make([]types.BlockID, 0, s.LayerSize)
+
+	//filter double ids
+	for _, bid := range out.([]types.BlockID) {
+		if _, exists := idSet[bid]; !exists {
+			idSet[bid] = struct{}{}
+			ids = append(ids, bid)
+		}
+	}
+
+	return ids, nil
+}
+
+type peerHashPair struct {
+	peer p2p.Peer
+	hash []byte
+}
+
+func (s *Syncer) fetchLayerHashes(lyr types.LayerID) (map[string]p2p.Peer, error) {
+	// get layer hash from each peer
+	wrk, output := NewPeersWorker(s, s.GetPeers(), &sync.Once{}, HashReqFactory(lyr))
+	go wrk.Work()
+	m := make(map[string]p2p.Peer)
+	for out := range output {
+		pair := out.(*peerHashPair)
+		if pair == nil { //do nothing on close channel
+			continue
+		}
+		m[string(pair.hash)] = pair.peer
```

Comment:
calling string will cause non readable characters in dict, are you sure this is what you want?

Reply:
yes